File: backend/app/core/upgrades/stat_modifiers/polygon_upgrade.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class PolygonUpgrade(UpgradeImplementation):
    id = 'polygon_upgrade'
    name = 'Evolution'
    description = '+1 polygon side'
    rarity = 'legendary'
    type = 'stat_modifier'
    cost = 60

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 1
        appliers.modifiers.add_multiplier('player', 'polygonSides', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/polygon_upgrade.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const PolygonUpgradeDef: UpgradeDef = {
  id: "polygon_upgrade",
  name: "Evolution",
  description: "+1 polygon side",
  rarity: RarityID.Legendary,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 60,
  targetClass: UpgradeTargetID.Player,
  fieldInTargetClass: UpgradeStatID.PolygonSides,
  value: 1,
  stackable: true,
  maxStacks: 9,
}

export class PolygonUpgrade extends Upgrade {}
